stan/math/rev/core/vari.hpp

```diff
@@ -179,11 +165,375 @@ class vari_value<T, std::enable_if_t<std::is_floating_point<T>::value>>
+   * adjoint is initialized to zero.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  explicit vari_value(S&& x)
+      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        val_(eigen_map(val_mem_, x.rows(), x.cols()) = x),
+        adj_(eigen_map(adj_mem_, x.rows(), x.cols()).setZero()) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+  /**
+   * Construct an dense Eigen variable implementation from a value. The
+   *  adjoint is initialized to zero and if `stacked` is `false` this vari
+   *  will be not be put on the var_stack. Instead it will only be put on
+   *  a stack to keep track of whether the adjoint needs to be set to zero.
+   *
+   * All constructed variables are added to a stack.  Variables
+   *  should be constructed before variables on which they depend
+   *  to insure proper partial derivative propagation.  During
+   *  derivative propagation, the chain() method of each variable
+   *  will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   * @param stacked If false will put this this vari on the nochain stack so
+   * that its `chain()` method is not called.
+   */
+  template <typename S, require_convertible_t<S&, value_type>* = nullptr>
+  vari_value(S&& x, bool stacked)
+      : val_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        adj_mem_(ChainableStack::instance_->memalloc_.alloc_array<eigen_scalar>(
+            x.size())),
+        val_(eigen_map(val_mem_, x.rows(), x.cols()) = x),
+        adj_(eigen_map(adj_mem_, x.rows(), x.cols()).setZero()) {
+    if (stacked) {
+      ChainableStack::instance_->var_stack_.push_back(this);
+    } else {
+      ChainableStack::instance_->var_nochain_stack_.push_back(this);
+    }
+  }
+
+  /**
+   * Return the number of rows for this class's `val_` member
+   */
+  const Eigen::Index rows() const { return val_.rows(); }
+  /**
+   * Return the number of columns for this class's `val_` member
+   */
+  const Eigen::Index cols() const { return val_.rows(); }
+  /**
+   * Return the size of this class's `val_` member
+   */
+  const Eigen::Index size() const { return val_.size(); }
+
+  virtual void chain() {}
+  /**
```

Comment:
I don't think init_dependent makes sense here (there's no single obvious thing to initialize it to -- reverse mode differentiation gets us gradients of one thing -- not gradients of a bunch of things [every element in a matrix]). I'd like to just get rid of it if possible.

Reply:
I'm not sure I'm following here. In `grad()` we call `vi->init_dependent();` to set the adjoint to ones. Would should this start with instead?

Reply:
Here's an example: https://github.com/stan-dev/math/blob/7811368763e51bdc65a263a186e41656a63e6cec/stan/math/rev/functor/cvodes_integrator.hpp#L276

There's a for loop that fills up a bunch of adjoints and then I call `grad()` to do the reverse mode autodiff. `grad()` has been modified to not require a vari, and with no vari it does no `init_dependent` call.

This is a situation where the `init_dependent` isn't obvious, so we don't have it.

Reply:
> Would should this start with instead?

So for your example you have what the adjoint is being filled in with which is what I'm asking here. 

>  reverse mode differentiation gets us gradients of one thing -- not gradients of a bunch of things [every element in a matrix]). I'd like to just get rid of it if possible.

Isn't the point of the static stuff to take the gradients of every element of a matrix (aka the whole matrix) at once?

Reply:
> Isn't the point of the static stuff to take the gradients of every element of a matrix (aka the whole matrix) at once?

Reverse mode autodiff is effectively giving us gradients of a function with one output.

When we have a function that outputs a vector, when we set init_dependent to all ones, we are effectively taking the gradient of that vector function dotted with a vector of ones. It's useful to be able to dot the function with all sorts of vectors -- there's nothing special about the ones (and no reason to use that as the default).

Something like that at least.

Reply:
Okay I think I'm seeing what you mean now. What if we just had a grad function that we passed in a lambda to like

```cpp
template <typename Vari, typename Init>
static void grad(Vari* vi, Init&& init_func) {
  init_func(vi);
  size_t end = ChainableStack::instance_->var_stack_.size();
  size_t beginning = empty_nested() ? 0 : end - nested_size();
  for (size_t i = end; i-- > beginning;) {
    ChainableStack::instance_->var_stack_[i]->chain();
  }
}
```

Reply:
I am fond of:

```
static void grad(Vari* vi = nullptr) {
  if(vi) {
    vi->init_dependent(); // just make init_dependent throw for vari_value<Eigen> types
  }
  
  ...
}
```

Reply:
Why does it need to throw? I'm not a big fan of that style personally I'd rather keep initialization as part of taking the gradient (aka passing the initiliazer as a lambda). Otherwise sometimes the gradient is ready to go here and other times you need to read the code around the call site to get the context on what this function is actually going to do.

Also again what would we do for default initlaization for `vari<Eigen>`. Would each one need it's own initializer? I understand it doesn't need to be set to ones, but it does need to be set to not zero values right?

Reply:
I think this is the only thing we are waiting on? imo this sounds like it's optional and we can have a separate PR for initializing with whatever we want in the future